controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -632,11 +632,29 @@ public void initializeStreamWriters(final EventStreamClientFactory clientFactory
     @VisibleForTesting
     <T> CompletableFuture<T> addIndexAndSubmitTask(ControllerEvent event, Supplier<CompletableFuture<T>> futureSupplier) {
         String id = UUID.randomUUID().toString();
+        // We first add index and then call the metadata update.
+        // In case, while trying to perform a metadata update, if we get connection exception or write conflict exception 
+        // (which can also occur if we had retried on a store exception), we will still post the event because we dont know 
+        // if our update succeeded or not. Posting the event is harmless. If the update had succeeded, then the event will 
+        // be used for processing. If the update had failed, the event will be discarded. We will throw the exception 
+        // that we thrown from running futureSupplier.
+        AtomicReference<Throwable> toThrow = new AtomicReference<>();
         return streamMetadataStore.addRequestToIndex(context.getHostId(), id, event)
-                           .thenCompose(v -> futureSupplier.get())
+                           .thenCompose(v -> Futures.exceptionallyExpecting(futureSupplier.get(), 
+                                   e -> {
+                                       toThrow.set(e);
+                                       return Exceptions.unwrap(e) instanceof StoreException.StoreConnectionException ||
+                                               Exceptions.unwrap(e) instanceof StoreException.WriteConflictException;
+                                   }, null))
                            .thenCompose(t -> RetryHelper.withIndefiniteRetriesAsync(() -> writeEvent(event), e -> { }, executor)
                                                         .thenCompose(v -> streamMetadataStore.removeTaskFromIndex(context.getHostId(), id))
-                                                        .thenApply(v -> t));
+                                                        .thenApply(v -> {
+                                                            if (toThrow.get() != null) {
```

Comment:
if `streamMetadataStore.removeTaskFromIndex(...)` throws an exception we would never know that `futureSupplier.get()` faced an exception right?

Reply:
I don’t follow. It is inside a completionStage callback so our future will complete exceptionally. What is your concern exactly?

Reply:
ok, got it.